google-cloud-spanner/lib/google/cloud/spanner/client.rb

```diff
@@ -728,7 +731,11 @@ def transaction deadline: 120, &block
           start_time = current_time
 
           @pool.with_transaction do |tx|
+            unless Thread.current[:transaction_id].nil?
+              fail "Nested transactions are not allowed"
+            end
             begin
+              Thread.current[:transaction_id] = tx.transaction_id
```

Comment:
I don't think that is necessary. We only care if a value is set or not. The `begin` block can be retried with a new transaction value, so it is cleaner to have the begin block set the value. When a value is present an error is raised _before_ the begin block begins.